packages/tools/src/eventListeners/touch/touchStartListener.ts

```diff
@@ -290,126 +335,128 @@ function _onTouchDrag(evt: TouchEvent) {
+    currentPointsList,
+    startPoints: getMeanTouchPoints(state.startPointsList),
+    lastPoints: getMeanTouchPoints(lastPointsList),
+    currentPoints: getMeanTouchPoints(currentPointsList),
+    deltaPoints,
+    deltaDistance,
+  };
+
+  triggerEventCallback(eventDetail.element, TOUCH_END, eventDetail);
+  _checkTouchTap(evt);
 
+  // reset to default state
+  state = JSON.parse(JSON.stringify(defaultState));
+  document.removeEventListener('touchmove', _onTouchDrag);
+  document.removeEventListener('touchend', _onTouchEnd);
+}
+
+function _checkTouchTap(evt: TouchEvent): void {
+  const currentTime = new Date().getTime();
+  const startTime = state.startTime.getTime();
+  if (currentTime - startTime >= tapState.tapWindow) return;
+
+  // first tap, initialize the state
+  if (tapState.taps === 0) {
+    tapState.element = state.element;
+    tapState.renderingEngineId = state.renderingEngineId;
+    tapState.viewportId = state.viewportId;
+    tapState.startPointsList = state.startPointsList;
+    tapState.startTime = new Date();
+  }
+
+  // check that taps are within interval
+  if (currentTime - tapState.startTime.getTime() >= tapState.tapWindow) return;
+
+  // subsequent tap is on a different element
   if (
-    state.taps > 0 &&
-    getDeltaDistance(currentPointsList, lastPointsList).canvas <
-      state.tapMaxDistance
+    tapState.taps > 0 &&
+    !(
+      tapState.element == state.element &&
+      tapState.renderingEngineId == state.renderingEngineId &&
+      tapState.viewportId == state.viewportId
+    )
   ) {
-    state.taps = state.taps + 1;
+    return;
   }
 
-  if (state.taps === 0) {
-    state.taps = state.taps + 1;
-  }
+  const currentPointsList = getTouchEventPoints(evt, tapState.element);
+  const distanceFromStart = getDeltaDistance(
+    currentPointsList,
+    tapState.startPointsList
+  ).canvas;
 
-  state.lastPointsList = copyPointsList(currentPointsList); // Update the last points
+  if (distanceFromStart < tapState.tapMaxDistance) {
+    clearTimeout(tapState.tapTimeout);
+    tapState.startTime = new Date();
+    tapState.taps += 1;
+  }
 
-  state.tapTimeout = setTimeout(() => {
+  tapState.tapTimeout = setTimeout(() => {
     const eventDetail: EventTypes.TouchTapEventDetail = {
       event: evt,
       eventName: TOUCH_TAP,
-      element: state.element,
-      renderingEngineId: state.renderingEngineId,
-      viewportId: state.viewportId,
+      element: tapState.element,
+      renderingEngineId: tapState.renderingEngineId,
+      viewportId: tapState.viewportId,
       camera: {},
       currentPointsList,
       currentPoints: getMeanTouchPoints(currentPointsList),
-      taps: state.taps,
+      taps: tapState.taps,
     };
     triggerEventCallback(eventDetail.element, TOUCH_TAP, eventDetail);
-    state = JSON.parse(JSON.stringify(defaultState));
-  }, state.touchStartDelay - 10);
+    tapState = JSON.parse(JSON.stringify(defaultTapState));
+  }, tapState.tapWindow);
```

Comment:
So the reason for the window is to aggregate multiple taps. If the user taps, there will always be a delay of tapWindowMs before the actual tap fires.

In the case the user taps again within the TapWindow, the setTimeout would clear and wait for another tapWindowMs to pass before firing an event that *2* taps were heard.

The alternative approach to this is that we fire events every time a tap happens. This means a triple tap would actually be 3 events. This would prevent lag, but the tap events can get noisy. The windowing logic would then need to be pushed down for any listener of the tap event to set their own windows + timeouts.

Tap(1)
Tap(2)
Tap(3)

Reply:
Yes I understand why the aggregation and NO, we do NOT want the alternative approach you suggested.

I apologize because I think I misread the code before. However having both lines 379 and 391 are a bit confusing to me.

Line 379 together with the `setTimeout` on line 417 sort of suggests to me that a single tap has to fall within 300ms (i.e. tapState.tapWindow), double tap has to fall within 600ms, etc. However line 391 to me reads, for example, if the second tap happens 300ms (say 400ms) after the first tap, then do not do the double tap (i.e. return). JSYK I verified this via debugging whereby the debugger showed me that I did two taps within 351ms and I did NOT get a double tap. I honestly do not think we need both checks, but I could be wrong.